iceoryx_utils/include/iceoryx_utils/internal/lockfree_queue/index_queue.inl

```diff
@@ -0,0 +1,274 @@
+namespace iox
+{
+template <uint64_t Capacity, typename ValueType>
+IndexQueue<Capacity, ValueType>::IndexQueue(ConstructEmpty_t)
+    : m_readPosition(Index(Capacity))
+    , m_writePosition(Index(Capacity))
+{
+}
+
+template <uint64_t Capacity, typename ValueType>
+IndexQueue<Capacity, ValueType>::IndexQueue(ConstructFull_t)
+    : m_readPosition(Index(0))
+    , m_writePosition(Index(Capacity))
+{
+    for (uint64_t i = 0; i < Capacity; ++i)
+    {
+        m_cells[i].store(Index(i));
+    }
+}
+
+template <uint64_t Capacity, typename ValueType>
+constexpr uint64_t IndexQueue<Capacity, ValueType>::capacity()
+{
+    return Capacity;
+}
+
+template <uint64_t Capacity, typename ValueType>
+void IndexQueue<Capacity, ValueType>::push(ValueType index)
+{
+    // we need the CAS loop here since we may fail due to concurrent push operations
+    // note that we are always able to succeed to publish since we have
+    // enough capacity for all unique indices used
+
+    // case analyis
+    // (1) loaded value is exactly one cycle behind:
+    //     value is from the last cycle
+    //     we can try to publish
+    // (2) loaded value has the same cycle:
+    //     some other push has published but not updated the write position
+    //     help updating the write position
+    // (3) loaded value is more than one cycle behind:
+    //     this should only happen due to wrap around when push is interrupted for a long time
+    //     reload write position and try again
+    //     note that a complete wraparound can lead to a false detection of 1) (ABA problem)
+    //     but this is very unlikely with e.g. a 64 bit value type
+    // (4) loaded value is some cycle ahead:
+    //     write position is outdated, there must have been other pushes concurrently
+    //     reload write position and try again
+
+    constexpr bool notPublished = true;
+    auto writePosition = loadNextWritePosition();
+    do
+    {
+        auto oldValue = loadValueAt(writePosition);
+
+        auto cellIsFree = oldValue.isOneCycleBehind(writePosition);
+
+        if (cellIsFree)
+        {
+            // case (1)
+            Index newValue(index, writePosition.getCycle());
+
+            // if publish fails, another thread has published before us
+            if (tryToPublishAt(writePosition, oldValue, newValue))
+            {
+                break;
+            }
+        }
+
+        // even if we are not able to publish, we check whether some other push has already updated the writePosition
+        // before trying again to publish
+        auto writePositionRequiresUpdate = oldValue.getCycle() == writePosition.getCycle();
+
+        if (writePositionRequiresUpdate)
+        {
+            // case (2)
+            // the writePosition was not updated yet by another push but the value was already written
+            // help with the update
+            updateNextWritePosition(writePosition);
```

Comment:
Why does the concurrent push need help ?

Reply:
This probably requires more explanation in a comment, I think we cannot have code easily expressing this on its own (the ideal world). I will add this explanation, but the main reason is this: to complete a push operation.

Reply:
It should also e noted that if no push comes to help, it is not a problem since this means there are no pushes needing the update. Pops are unaffected by the incomplete operation, they do not require writePosition (well popIfFull does, but that is more complicated, this one would fail which would be correct, because technically the incomplete push means the queue cannot be full...).